pkg/kubernetes/controllerutil.go

```diff
@@ -0,0 +1,138 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// OperationResultUpdatedStatus means that an existing resource and its status is updated
+	OperationResultUpdatedStatus controllerutil.OperationResult = "updatedStatus"
+	// OperationResultUpdatedStatusOnly means that only an existing status is updated
+	OperationResultUpdatedStatusOnly controllerutil.OperationResult = "updatedStatusOnly"
+)
+
+// CreateOrPatch creates or patches the given object in the Kubernetes
+// cluster. The object's desired state must be reconciled with the before
+// state inside the passed in callback MutateFn.
+//
+// The MutateFn is called regardless of creating or updating an object.
+//
+// It returns the executed operation and an error.
+//
+// Ported from https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go
+func CreateOrPatch(ctx context.Context, c client.Client, obj runtime.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	key, err := client.ObjectKeyFromObject(obj)
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !errors.IsNotFound(err) {
+			return controllerutil.OperationResultNone, err
+		}
+		if f != nil {
+			if err := mutate(f, key, obj); err != nil {
+				return controllerutil.OperationResultNone, err
+			}
+		}
+		if err := c.Create(ctx, obj); err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+		return controllerutil.OperationResultCreated, nil
+	}
+
+	if err := c.Get(ctx, key, obj); err != nil {
+		return controllerutil.OperationResultNone, err
+	}
```

Comment:
This differs from the original version. Do we need to reload the object?

Reply:
It seems to Get always reload object and than mutate it with function. I don't know why we need to reload object every time.

Reply:
I can't understand, but I was probably unclear.
I mean why do we need an additional c.Get() call, which is not present in the original implementation ?
https://github.com/kubernetes-sigs/controller-runtime/blob/dea6dee6c3e1ae2a2c0555f37ab28266105a1206/pkg/controller/controllerutil/controllerutil.go#L249

Reply:
Oh, @rigazilla, are you talking about our first c.Get inside Infinspan controller? The reason why we need it, it's an non standard NotFound handle to prevent Create it in case of delete

Reply:
I'm not sure, I'm referring to [these lines of code](https://github.com/infinispan/infinispan-operator/blob/c7da3615fb724b0892c619f77187f5cbf04c5059/pkg/kubernetes/controllerutil.go#L51-L53) of the `CreateOrPatch` in this PR which introduces an additional c.Get() which is not present in [the original version](https://github.com/kubernetes-sigs/controller-runtime/blob/1d63e429f706bd9ad81a1e47642745ed20e76bd1/pkg/controller/controllerutil/controllerutil.go#L250)

Reply:
Oh, @rigazilla thanks a lot
This was missing from some tests. I will fix and validate 2.0.x backport.also

Reply:
Fixed. Please review